### ISSUE TITLE
Bug in visualizing pvtu files

### ISSUE THREAD

**Issue by aarograh** [visit-dav/live-customer-response#267]
I believe I've identified a bug in VisIt related to reading multiple .vtu files via a .pvtu file.

I have the following results of a simulation:
```
fsr_4a-2d/
    4a-2d_core_FSRmesh_STATE_0001.pvtu
    DOMAIN_0/
        4a-2d_core_FSRmesh_STATE_0001_0.vtu
    DOMAIN_1/
        4a-2d_core_FSRmesh_STATE_0001_1.vtu
    DOMAIN_2/
        4a-2d_core_FSRmesh_STATE_0001_2.vtu
    DOMAIN_3/
        4a-2d_core_FSRmesh_STATE_0001_3.vtu
    DOMAIN_4/
        4a-2d_core_FSRmesh_STATE_0001_4.vtu
    DOMAIN_5/
        4a-2d_core_FSRmesh_STATE_0001_5.vtu
    DOMAIN_6/
        4a-2d_core_FSRmesh_STATE_0001_6.vtu
    DOMAIN_7/
        4a-2d_core_FSRmesh_STATE_0001_7.vtu
    DOMAIN_8/
        4a-2d_core_FSRmesh_STATE_0001_8.vtu
```

The contents of the .pvtu file:
```
<?xml version="1.0"?>
<VTKFile type="PUnstructuredGrid" version="0.1" byte_order="LittleEndian">
  <PUnstructuredGrid GhostLevel="0">
    <PPoints>
      <PDataArray type="Float32" NumberOfComponents="3"/>
    </PPoints>
    <PCells>
      <PDataArray type="Int32" Name="connectivity"/>
      <PDataArray type="Int32" Name="offsets"/>
      <PDataArray type="UInt8" Name="types"/>
    </PCells>
    <PCellData Scalars="Data">
      <PDataArray type="Int32" Name="vcell_Index"/>
      <PDataArray type="Int32" Name="mesh_Index"/>
      <PDataArray type="Int32" Name="material"/>
    </PCellData>
    <Piece Source="DOMAIN_0/4a-2d_core_FSRmesh_STATE_0001_0.vtu"/>
    <Piece Source="DOMAIN_1/4a-2d_core_FSRmesh_STATE_0001_1.vtu"/>
    <Piece Source="DOMAIN_2/4a-2d_core_FSRmesh_STATE_0001_2.vtu"/>
    <Piece Source="DOMAIN_3/4a-2d_core_FSRmesh_STATE_0001_3.vtu"/>
    <Piece Source="DOMAIN_4/4a-2d_core_FSRmesh_STATE_0001_4.vtu"/>
    <Piece Source="DOMAIN_5/4a-2d_core_FSRmesh_STATE_0001_5.vtu"/>
    <Piece Source="DOMAIN_6/4a-2d_core_FSRmesh_STATE_0001_6.vtu"/>
    <Piece Source="DOMAIN_7/4a-2d_core_FSRmesh_STATE_0001_7.vtu"/>
    <Piece Source="DOMAIN_8/4a-2d_core_FSRmesh_STATE_0001_8.vtu"/>
  </PUnstructuredGrid>
</VTKFile>
```

The structure of `fsr_4a-2d/DOMAIN_0/4a-2d_core_FSRmesh_STATE_0001_0.vtu` is as follows (skipping actual values since the files are rather large):
```
<?xml version="1.0"?>
<VTKFile type="UnstructuredGrid" version="0.1" byte_order="LittleEndian">
  <UnstructuredGrid>
    <Piece NumberOfPoints="24177" NumberOfCells="12710">
      <Points>
        <DataArray type="Float32" NumberOfComponents="3" format="ascii">
        </DataArray>
      </Points>
      <Cells>
        <DataArray type="Int32" Name="connectivity" format="ascii">
        </DataArray>
        <DataArray type="Int32" Name="offsets" format="ascii">
        </DataArray>
        <DataArray type="UInt8" Name="types" format="ascii">
        </DataArray>
      </Cells>
      <CellData>
        <DataArray type="Int32" Name="vcell_Index" format="ascii">
        </DataArray>
        <DataArray type="Int32" Name="mesh_Index" format="ascii">
        </DataArray>
        <DataArray type="Int32" Name="material" format="ascii">
        </DataArray>
      </CellData>
    </Piece>
  </UnstructuredGrid>
</VTKFile>
```

The issue arises with the `material` dataset.  Domain 0 has material IDs 1, 12, 22, and 28.  Domain 2 has material IDs 1, 12, 22, 25.  When I open the .pvtu file and attempt to do a Filled Boundary plot of the materials, I get the following error:

![image](https://user-images.githubusercontent.com/9356636/118144542-92ad3780-b3da-11eb-8312-e972c14b4b88.png)

If I switch Domain 0 and Domain 2 (swapping 0 and 2 in the names of the directories and the files), then I get the exact same error, except that it now complains about material 28 instead of material 25.

Going back to the original case, if I manually open the Domain 0 .vtu file and change a single material ID to 25, then I can open the .pvtu file in Visit and everything looks as it should (except for the one small region that has material 25 instead of 28, but obviously that's because I changed it manually).

It seems that VisIt is somehow caching a list of allowed materials based only on Domain 0 instead of looking at the possible material IDs in every domain.  This type of behavior is pretty consistent with other domains and IDs.

I tested this with VisIt 3.1.2 and 3.2.  Both versions displayed the exact same behavior.

**Comment by brugger1** [comment]
Hi,

You are correct in your assessment that VisIt is looking at the first piece to get a complete list of the material numbers for all the pieces. It does this so that it doesn't need to read all the blocks to get the list of material numbers.

Fortunately, you can add another piece of data to each of your vtu files (probably the first one would be sufficient) that lists all the materials in the pieces. You can do this by adding field data to the unstructured grid:

```
  <UnstructuredGrid>
    <FieldData>
      <DataArray type="Int32" Name="MaterialIds" NumberOfTuples="5" format="ascii">
      1 12 22 25 28
      </DataArray>
    </FieldData>
```

Here is a complete example of 4 hexahedron in 2 pieces. The pvtu file:

```
<?xml version="1.0"?>
<VTKFile type="PUnstructuredGrid" version="0.1" byte_order="LittleEndian">
  <PUnstructuredGrid GhostLevel="0">
    <PPoints>
      <PDataArray type="Float32" NumberOfComponents="3"/>
    </PPoints>
    <PCells>
      <PDataArray type="Int32" Name="connectivity"/>
      <PDataArray type="Int32" Name="offsets"/>
      <PDataArray type="UInt8" Name="types"/>
    </PCells>
    <PCellData Scalars="Data">
      <PDataArray type="Int32" Name="material"/>
    </PCellData>
    <Piece Source="test1.vtu"/>
    <Piece Source="test2.vtu"/>
  </PUnstructuredGrid>
</VTKFile>
```

test1.vtu:

```
<?xml version="1.0"?>
<VTKFile type="UnstructuredGrid" version="0.1" byte_order="LittleEndian">
  <UnstructuredGrid>
    <FieldData>
      <DataArray type="Int32" Name="MaterialIds" NumberOfTuples="3" format="ascii">
      1 2 3
      </DataArray>
    </FieldData>
    <Piece NumberOfPoints="12" NumberOfCells="2">
      <Points>
        <DataArray type="Float32" NumberOfComponents="3" format="ascii">
        0. 0. 0.
        0. 1. 0.
        0. 2. 0.
        1. 0. 0.
        1. 1. 0.
        1. 2. 0.
        0. 0. 1.
        0. 1. 1.
        0. 2. 1.
        1. 0. 1.
        1. 1. 1.
        1. 2. 1.
        </DataArray>
      </Points>
      <Cells>
        <DataArray type="Int32" Name="connectivity" format="ascii">
        0 1 4 3 6 7 10 9
        1 2 5 4 7 8 11 10
        </DataArray>
        <DataArray type="Int32" Name="offsets" format="ascii">
        8 16
        </DataArray>
        <DataArray type="UInt8" Name="types" format="ascii">
        12 12
        </DataArray>
      </Cells>
      <CellData>
        <DataArray type="Int32" Name="material" format="ascii">
        1 2
        </DataArray>
      </CellData>
    </Piece>
  </UnstructuredGrid>
</VTKFile>
```

test2.vtu:
```
<?xml version="1.0"?>
<VTKFile type="UnstructuredGrid" version="0.1" byte_order="LittleEndian">
  <UnstructuredGrid>
    <FieldData>
      <DataArray type="Int32" Name="MaterialIds" NumberOfTuples="3" format="ascii">
      1 2 3
      </DataArray>
    </FieldData>
    <Piece NumberOfPoints="12" NumberOfCells="2">
      <Points>
        <DataArray type="Float32" NumberOfComponents="3" format="ascii">
        1. 0. 0.
        1. 1. 0.
        1. 2. 0.
        2. 0. 0.
        2. 1. 0.
        2. 2. 0.
        1. 0. 1.
        1. 1. 1.
        1. 2. 1.
        2. 0. 1.
        2. 1. 1.
        2. 2. 1.
        </DataArray>
      </Points>
      <Cells>
        <DataArray type="Int32" Name="connectivity" format="ascii">
        0 1 4 3 6 7 10 9
        1 2 5 4 7 8 11 10
        </DataArray>
        <DataArray type="Int32" Name="offsets" format="ascii">
        8 16
        </DataArray>
        <DataArray type="UInt8" Name="types" format="ascii">
        12 12
        </DataArray>
      </Cells>
      <CellData>
        <DataArray type="Int32" Name="material" format="ascii">
        2 3
        </DataArray>
      </CellData>
    </Piece>
  </UnstructuredGrid>
</VTKFile>
```

**Comment by aarograh** [comment]
@brugger1 Thanks for the information!  I've implemented it in my code and it works just like you said.  Glad to know it's not a bug in VisIt but rather something I can fix on my end.

Do you happen to know of VTK/VTU documentation that gives this level of detail?  I looked around several different sources trying to resolve this issue and never found any discussion of field data.  I maybe have just been looking in the wrong places though.

**Comment by brugger1** [comment]
@aarograh Unfortunately this isn't documented anywhere. The use of MaterialIds is a VisIt convention and I had to look at the source code for the VTK reader to determine this. As far as documentation on VTU and PVTU files, I haven't been able to find much on the format of those files either. I usually have to figure it out from examples I've found. I've submitted issue https://github.com/visit-dav/visit/issues/5706 to add documentation about VTK files.
